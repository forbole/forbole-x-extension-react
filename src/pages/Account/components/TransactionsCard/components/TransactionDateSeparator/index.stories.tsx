import React from 'react';

import { ComponentStory, ComponentMeta } from '@storybook/react';

import TransactionDateSeparator from './index';

type CompType = typeof TransactionDateSeparator;

export default {
  title: 'Account/TransactionDateSeparator',
  component: TransactionDateSeparator,
} as ComponentMeta<CompType>;

export const Today: ComponentStory<CompType> = () => (
  <TransactionDateSeparator daysFromPresent={0} />
);

export const Yesterday: ComponentStory<CompType> = () => (
  <TransactionDateSeparator daysFromPresent={1} />
);

export const XDaysAgo: ComponentStory<CompType> = () => (
  <>
    <TransactionDateSeparator daysFromPresent={3} />
    <TransactionDateSeparator daysFromPresent={6} />
  </>
);

export const AWeekAgo: ComponentStory<CompType> = () => (
  <TransactionDateSeparator daysFromPresent={7} />
);

export const MoreThanAWeek: ComponentStory<CompType> = () => (
  <TransactionDateSeparator daysFromPresent={1000} />
);
