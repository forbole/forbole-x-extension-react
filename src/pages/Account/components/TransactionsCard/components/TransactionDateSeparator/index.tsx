import React from 'react';
import { Typography } from '@mui/material';
import { useTranslation } from 'react-i18next';
import { format, subDays } from 'date-fns';

type Props = {
  daysFromPresent: number;
};

/**
 * A separator component that is used to separate transaction histories by date.
 * Transaction histories are ordered in the steps outlined below.
 */
const TransactionDateSeparator = ({ daysFromPresent }: Props) => {
  const { t } = useTranslation('account');

  /**
   * Time steps that will be rendered:
   * Today
   * Yesterday
   * xx days ago (up to 6)
   * a week ago
   * anything more than 7 days: should show the date in MMM dd yyyy format
   */

  const dateContent = React.useMemo(() => {
    if (daysFromPresent === 0) {
      return t('transactions.rows.today');
    }
    if (daysFromPresent === 1) {
      return t('transactions.rows.yesterday');
    }
    if (daysFromPresent < 7) {
      return t('transactions.rows.xDaysAgo', { days: daysFromPresent });
    }
    if (daysFromPresent === 7) {
      return t('transactions.rows.aWeekAgo');
    }
    return format(subDays(new Date(), daysFromPresent), 'MMM dd yyyy');
  }, [daysFromPresent]);

  return (
    <Typography padding={2} variant="body1" color="text.secondary">
      {dateContent}
    </Typography>
  );
};

export default TransactionDateSeparator;
