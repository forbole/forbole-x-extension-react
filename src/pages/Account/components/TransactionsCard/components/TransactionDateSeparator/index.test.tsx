import React from 'react';
import { render } from '@testing-library/react';
import TransactionDateSeparator from './index';

describe('components/TransactionDateSeparator', () => {
  it('renders (today)', () => {
    const t = render(<TransactionDateSeparator daysFromPresent={0} />);

    expect(t).toMatchSnapshot();
  });

  it('renders (yesterday)', () => {
    const t = render(<TransactionDateSeparator daysFromPresent={0} />);

    expect(t).toMatchSnapshot();
  });

  it('renders (x days ago)', () => {
    const t = render(
      <>
        <TransactionDateSeparator daysFromPresent={2} />
        <TransactionDateSeparator daysFromPresent={3} />
        <TransactionDateSeparator daysFromPresent={4} />
        <TransactionDateSeparator daysFromPresent={5} />
        <TransactionDateSeparator daysFromPresent={6} />
      </>
    );

    expect(t).toMatchSnapshot();
  });

  it('renders (a week ago)', () => {
    const t = render(<TransactionDateSeparator daysFromPresent={7} />);

    expect(t).toMatchSnapshot();
  });

  it('renders (date in MM dd yyyy format)', () => {
    jest.useFakeTimers().setSystemTime(new Date('2022-01-01'));
    const t = render(<TransactionDateSeparator daysFromPresent={1000} />);

    expect(t).toMatchSnapshot();

    jest.useRealTimers();
  });
});
