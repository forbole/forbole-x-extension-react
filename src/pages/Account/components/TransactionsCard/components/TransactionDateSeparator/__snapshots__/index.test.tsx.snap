// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/TransactionDateSeparator renders (a week ago) 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <p
        class="MuiTypography-root MuiTypography-body1 css-masy5k-MuiTypography-root"
      >
        transactions.rows.aWeekAgo
      </p>
    </div>
  </body>,
  "container": <div>
    <p
      class="MuiTypography-root MuiTypography-body1 css-masy5k-MuiTypography-root"
    >
      transactions.rows.aWeekAgo
    </p>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`components/TransactionDateSeparator renders (date in MM dd yyyy format) 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <p
        class="MuiTypography-root MuiTypography-body1 css-masy5k-MuiTypography-root"
      >
        Sep 10 2019
      </p>
    </div>
  </body>,
  "container": <div>
    <p
      class="MuiTypography-root MuiTypography-body1 css-masy5k-MuiTypography-root"
    >
      Sep 10 2019
    </p>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`components/TransactionDateSeparator renders (today) 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <p
        class="MuiTypography-root MuiTypography-body1 css-masy5k-MuiTypography-root"
      >
        transactions.rows.today
      </p>
    </div>
  </body>,
  "container": <div>
    <p
      class="MuiTypography-root MuiTypography-body1 css-masy5k-MuiTypography-root"
    >
      transactions.rows.today
    </p>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`components/TransactionDateSeparator renders (x days ago) 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <p
        class="MuiTypography-root MuiTypography-body1 css-masy5k-MuiTypography-root"
      >
        transactions.rows.xDaysAgo
      </p>
      <p
        class="MuiTypography-root MuiTypography-body1 css-masy5k-MuiTypography-root"
      >
        transactions.rows.xDaysAgo
      </p>
      <p
        class="MuiTypography-root MuiTypography-body1 css-masy5k-MuiTypography-root"
      >
        transactions.rows.xDaysAgo
      </p>
      <p
        class="MuiTypography-root MuiTypography-body1 css-masy5k-MuiTypography-root"
      >
        transactions.rows.xDaysAgo
      </p>
      <p
        class="MuiTypography-root MuiTypography-body1 css-masy5k-MuiTypography-root"
      >
        transactions.rows.xDaysAgo
      </p>
    </div>
  </body>,
  "container": <div>
    <p
      class="MuiTypography-root MuiTypography-body1 css-masy5k-MuiTypography-root"
    >
      transactions.rows.xDaysAgo
    </p>
    <p
      class="MuiTypography-root MuiTypography-body1 css-masy5k-MuiTypography-root"
    >
      transactions.rows.xDaysAgo
    </p>
    <p
      class="MuiTypography-root MuiTypography-body1 css-masy5k-MuiTypography-root"
    >
      transactions.rows.xDaysAgo
    </p>
    <p
      class="MuiTypography-root MuiTypography-body1 css-masy5k-MuiTypography-root"
    >
      transactions.rows.xDaysAgo
    </p>
    <p
      class="MuiTypography-root MuiTypography-body1 css-masy5k-MuiTypography-root"
    >
      transactions.rows.xDaysAgo
    </p>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`components/TransactionDateSeparator renders (yesterday) 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <p
        class="MuiTypography-root MuiTypography-body1 css-masy5k-MuiTypography-root"
      >
        transactions.rows.today
      </p>
    </div>
  </body>,
  "container": <div>
    <p
      class="MuiTypography-root MuiTypography-body1 css-masy5k-MuiTypography-root"
    >
      transactions.rows.today
    </p>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
